### ISSUE TITLE
tree_map_only_ doesn't seem work as expected

### ISSUE THREAD

**Issue by wz337** [pytorch/pytorch#130709]
### 🐛 Describe the bug

I am trying to use `tree_map_only_` to initialize meta tensor in a state_dict. It seems `tree_map_only` (the out of place one) is working fine for me but `tree_map_only_ `(in place) doesn't seem to work. Just wonder whether this is an issue with `tree_map_only_` or I am not using it correctly. 

Below is a script for minimum repro. `state_dict` is still not initialized but `new_state_dict` is initialized.

```
class TestTreeMapOnly(TestCase):

    def _init_meta_tensor(self, value):
        device = getattr(value, "device", None)
        if device == torch.device("meta"):
            device = torch.device("cuda:0")
            if isinstance(value, torch.Tensor):
                tensor = torch.empty_like(value, device=device)
                return tensor
            else:
                raise RuntimeError(
                    f"Found unsupported type {type(value)} for meta device loading."
                )
        else:
            return value

    def test_tree_map_only(self):
        from torch.utils._pytree import tree_map_only_, tree_map_only

        with torch.device('meta'):
            model = torch.nn.Linear(10, 10)
            state_dict = model.state_dict()
        
        tree_map_only_(torch.Tensor, self._init_meta_tensor, state_dict)

        new_state_dict = tree_map_only(torch.Tensor, self._init_meta_tensor, state_dict)
        print(f"{state_dict=}\n{new_state_dict=}")
        """
        state_dict is still not initialized.
        state_dict=OrderedDict([('weight', tensor(..., device='meta', size=(10, 10))), ('bias', tensor(..., device='meta', size=(10,)))])
        
        new_state_dict is initialized.
        new_state_dict=OrderedDict([('weight', tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
                [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], device='cuda:0')), ('bias', tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], device='cuda:0'))])
        """


if __name__ == "__main__":
    run_tests()
```

cc. @fduwjj

### Versions

Nightly

**Comment by XuehaiPan** [comment]
You are creating a new tensor in `_init_meta_tensor()` rather than doing an in-place operation on the original tensor. FYI, `tensor.to('cuda')` is also not an inplace operation.

In the following example, `torch.zeros_like` is a non-inplace operation, and `torch.zero_` is an inplace operation.

```python
In [1]: import torch
   ...: import torch.nn as nn
   ...: import torch.nn.functional as F
   ...: import torch.utils._pytree as pytree

In [2]: state_dict = nn.Linear(3, 4).state_dict()

In [3]: state_dict
Out[3]: 
OrderedDict(
    {
        'weight': tensor([[ 0.5680, -0.0967, -0.0733],
                          [ 0.3857,  0.3045, -0.2402],
                          [-0.5311, -0.2645, -0.3222],
                          [ 0.5716, -0.4924,  0.5333]]),
        'bias': tensor([-0.1132, -0.3369,  0.2229,  0.2969])
    }
)

In [4]: pytree.tree_map_(torch.zeros_like, state_dict)
Out[4]: 
OrderedDict(
    {
        'weight': tensor([[ 0.5680, -0.0967, -0.0733],
                          [ 0.3857,  0.3045, -0.2402],
                          [-0.5311, -0.2645, -0.3222],
                          [ 0.5716, -0.4924,  0.5333]]),
        'bias': tensor([-0.1132, -0.3369,  0.2229,  0.2969])
    }
)

In [5]: state_dict
Out[5]: 
OrderedDict(
    {
        'weight': tensor([[ 0.5680, -0.0967, -0.0733],
                          [ 0.3857,  0.3045, -0.2402],
                          [-0.5311, -0.2645, -0.3222],
                          [ 0.5716, -0.4924,  0.5333]]),
        'bias': tensor([-0.1132, -0.3369,  0.2229,  0.2969])
    }
)

In [6]: pytree.tree_map_(torch.zero_, state_dict)
Out[6]: 
OrderedDict(
    {
        'weight': tensor([[0., 0., 0.],
                          [0., 0., 0.],
                          [0., 0., 0.],
                          [0., 0., 0.]]),
        'bias': tensor([0., 0., 0., 0.])
    }
)

In [7]: state_dict
Out[7]: 
OrderedDict(
    {
        'weight': tensor([[0., 0., 0.],
                          [0., 0., 0.],
                          [0., 0., 0.],
                          [0., 0., 0.]]),
        'bias': tensor([0., 0., 0., 0.])
    }
)
```

**Comment by XuehaiPan** [comment]
I do not find a way to create an on-device storage for a `meta` tensor. I think the solution should be using `state_dict + tree_map + load_state_dict`:

```python
mod.load_state_dict(pytree.tree_map_only(torch.Tensor, func, mod.state_dict()))
```

**Comment by wz337** [comment]
Understood. Thanks for the context.

Originally, I thought tree_map_only_ means container-based in place but looks like we require the element (the function used) is also in place. 

For more context, we are trying to let the original model/optim state dict to be updated in-place for distributed checkpoint. This is because for things like learning rate and primitive type, using the out of way would not really change these in place. So seems we need in-place for these non-tensor types but out-of-place for meta tensors.

**Comment by XuehaiPan** [comment]
> Originally, I thought tree_map_only_ means container-based in place but looks like we require the element (the function used) is also in place.

For some container types, we cannot make in place changes on the container. For example, `tuple`s, `namedtuple`s, and `torch.return_types.*`.

> This is because for things like learning rate and primitive type, using the out of way would not really change these in place. So seems we need in-place for these non-tensor types but out-of-place for meta tensors.

You can do:

```python
def func(x):
    if isinstance(x, torch.Tensor):
        if x.device.type = "meta":
            x = torch.zeros_like(x, device="cuda")
            return f(x)

        y = g(x)  # do something else
        return y

    # assume primitive types
    z = h(x)
    return z


mod.load_state_dict(pytree.tree_map(func, mod.state_dict()))
```

**Comment by wz337** [comment]
@XuehaiPan answered my question so I am closing the task.
